Cheats: Time Alteration (Skip Time, Faster Time, etc.)

Dumbledork01 opened the pull request:
**List of Cheats**
_(Note: All of these cheats have been put in their own sub-folder titled time...)_
* A slider in the cheats menu that adds to the rate of time of the current area. A similar cheat was available in a randomizer I played and I really liked having access to it if I didn't have the Sun's Song, so I figured I'd add it into here. 
* Cheats that change the time of day similar to morning, noon, evening, and night. (I added these from the debug menu because it'd be more convenient to have them in one place for new players.)
* Clock that now that tells you what the current time is.
* Button to reset the rate of time slider back to it's initial value.

**Demonstration Video**
Time Altering Cheats: https://youtu.be/CACLHKYsiQw

**Possible Issues**
* Rate of Time changer
An exception has been added for when the Sun's Song is active, I'm checking for what other exceptions should be added. If you know of any others, please leave a comment for me.

Dumbledork01 committed:
Create game.c

Based upon the area Link is in, the rate of time can be set to it's base value and can be incremented up 400 times (to the rate of the sun's song.) Useful for impatient people who don't have the sun's song and want to skulltula hunt or who don't want to wait around for dampe's digging tour.

Dumbledork01 committed:
Put Time Altering Cheats Together

Added in a slider to increase the rate of time. Put freeze time above it so they all are next to each other.

Dumbledork01 committed:
Exception for Sun's Song added

If the sun's song is active, the rate of time will not change even if the slider is on. I'm still looking into where else the rate of time is changed in the game.

Dumbledork01 committed:
Created a sub-menu for time cheats

Because I didn't want the cheats menu to be crowded with time cheats, I created a separate sub-menu for it like the infinite cheats. It has an easy to access location for changing the time of day to dawn, noon, evening, or midnight as well.

Dumbledork01 committed:
Added logic for setting time

Included logic to set time to morning, noon, evening, night, and to reset the slider for time.

Dumbledork01 committed:
Fixed typo

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into develop-zhora

Dumbledork01 committed:
Track time rate in any area

Now the rate of time slider will be able to display the current rate of time in any area.
Various Time Cheats
The slider will tell you the current time of day.
Added Skybox Updates
Now, when a person selects to change the time, the skybox will update.
Added a clock to cheats
Added a clock to tell what the time currently is.
Logic to display time rate
Added logic to show the current time rate on the time rate modifier slider.

aMannus commented:
Could you seperate the drawbridge and scarecrow features into their own PRs? They don't really have anything to do with each other, and makes reviewing/commenting for each feature easier. You'll have a better chance of stuff being merged in faster that way too.

Dumbledork01 commented:
@aMannus can do! I didn't want to flood the project with PRs since I have a few other things I want to add as well (skip frog song minigame, switch to adult or child link, skip zeldas lullaby, skip song of time) but I can do separate PRs for them. (I'll probably put all of the ocarina skips in the same PR though.)

aMannus commented:
> @aMannus can do! I didn't want to flood the project with PRs since I have a few other things I want to add as well (skip frog song minigame, switch to adult or child link, skip zeldas lullaby, skip song of time) but I can do separate PRs for them. (I'll probably put all of the ocarina skips in the same PR though.)

I'm not sure if those skips are worth having cheats for, but you're free to make PRs for them and have people look at it.

Dumbledork01 commented:
@aMannus I figure the ocarina skips could all be made into one enhancement called Quick Ocarina or something once I've got them made. I guess I'm just a lazy oot player haha.

aMannus commented:
Could you add text above the time speed slider indicating what the current value is? You can look at other sliders in the same file that do this already if you need an example.

Dumbledork01 commented:
@aMannus Should I put it above the slider, or inside of it similar to the Current Time Of Day slider I have beneath it? I can do either
(I added it onto the slider similar to how time is displayed in the video linked above, but I can change it if you want me to.)

aMannus commented:
![image](https://user-images.githubusercontent.com/4244591/184513882-728bf40e-4b6d-4927-abbc-4f1deba7cae1.png)
Like this

Dumbledork01 committed:
Added time rate tracking

Added text to show the current rate of time. Changes based upon the rate of time in a particular scene.

Dumbledork01 commented:
@aMannus Ok, the text should be working now!

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into develop-zhora

Dumbledork01 committed:
Added more spots for quick ocarina

I found that MANY spots are activated with this area of code (the triforces on the ground in temples, entrance to Darunia's room, Fairy Fountains, etc.) Exceptions not covered yet (that I know of) include Darunia himself, the Song of Storms grave, and possibly Time blocks. I will test more and see what else to add.

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into develop-zhora

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into develop-zhora

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into develop-zhora

Dumbledork01 committed:
Respawns link if time changes

If in an area where the time rate is 0, link is now respawned so the area can change to it's nighttime or daytime variant.

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into develop-zhora

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into develop-zhora

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into develop-zhora

Dumbledork01 committed:
Link returns to where he was prior to time change

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into develop-zhora

briaguya0 commented:
looks like this pr addresses https://github.com/HarbourMasters/Shipwright/issues/57, but also some other stuff

garrettjoecox commented:
@JakeEdvenson any interest in updating this PR? If not I am willing to take it on, I think these are improvements we want.

Dumbledork01 commented:
@garrettjoecox Yeah, if you're willing to update it, that'd be great. I haven't worked on this project in so long, I'd need to set up my IDE and whatnot to get it working again haha.

aMannus commented:
@garrettjoecox Is this still something you want to pick up and continue?

garrettjoecox commented:
> @garrettjoecox Is this still something you want to pick up and continue?

Yeap. On my todo list

briaguya0 commented:
i'm going to go ahead and close this one. a lot has changed since it was originally opened, and all of the code here will still be able to be referenced with the PR closed.

thank you so much for getting this PR out, sorry it fell through the cracks for so long.
